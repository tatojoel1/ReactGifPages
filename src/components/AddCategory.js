import React, { useState } from "react";
import { Form } from "react-bootstrap";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState(""); // ''

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    //console.log("handeSubmit", inputValue);
    if (inputValue.trim().length > 2) {
      setCategories((cats) => [inputValue, ...cats]);
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <Form.Group controlId="formBasicEmail">
        <p>{inputValue}</p>
        <Form.Label>Type in the box to seek some gifs for you</Form.Label>
        <input
          type="text"
          className="form-control"
          value={inputValue}
          onChange={handleInputChange}
          placeholder="Example...Tacos"
        />
      </Form.Group>
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
