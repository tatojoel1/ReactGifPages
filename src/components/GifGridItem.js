import React, { useState, useEffect, useRef } from "react";
import PropTypes from "prop-types";
import { Card, Button, Col, Tooltip, Overlay } from "react-bootstrap";
import { CopyToClipboard } from "react-copy-to-clipboard";

export const GifGridItem = ({ title, url }) => {
  const [copyText, setcopyText] = useState(false);
  const [show, setShow] = useState(false);
  const target = useRef(null);

  useEffect(() => {
    setTimeout(() => {
      setcopyText(false);
      setShow(false);
    }, 2000);
  }, [copyText]);

  return (
    <Col md={4} className="mb-4">
      <Card className="card animate__animated animate__fadeIn">
        <Card.Img variant="top" src={url} alt={title} />
        <Card.Body>
          <Card.Title className="card-title">
            <p>{title}</p>
            {/* {title.length > 0 ? title : "No title found"} */}
          </Card.Title>
        </Card.Body>
        <Card.Text>
          <h4>Link of gif: </h4>
          <a href={url} target="_blank" rel="noopener noreferrer">
            {url}
          </a>
        </Card.Text>
        {/* <Card.Text>{title.length > 0 ? "" : "Sorry, but the author doesn't put it some title in the gif"}</Card.Text> */}
        <Card.Footer>
          <CopyToClipboard text={url} onCopy={() => setcopyText(true)}>
            <Button
              className="copy-button rounded-pill"
              ref={target}
              onClick={() => setShow(!show)}
            >
              Copy
            </Button>
          </CopyToClipboard>
          <Overlay target={target.current} show={show} placement="right">
            {(props) => (
              <Tooltip id="overlay-example" {...props}>
                Copied!
              </Tooltip>
            )}
          </Overlay>
        </Card.Footer>
      </Card>
    </Col>
  );
};

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
