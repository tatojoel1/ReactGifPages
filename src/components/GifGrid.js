import React from "react";
import PropTypes from 'prop-types'
import { CardGroup, Row } from "react-bootstrap";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
  const { data, loading } = useFetchGifs(category);

  return (
    <>
      <CardGroup>
        <h3 className="animate__animated animate__fadeIn"> {category} </h3>

        {loading && <p className="animate__animated animate__flash">Loading</p>}

        <div className="card-grid">
          <Row md={4}>
            {data.map((img) => (
              <GifGridItem key={img.id} {...img} />
            ))}
          </Row>
        </div>
      </CardGroup>
    </>
  );
};

GifGrid.prototypes = {
  category: PropTypes.string.isRequired
}
