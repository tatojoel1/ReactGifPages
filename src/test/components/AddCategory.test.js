import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe("Prueba en el componente AddCategory", () => {
  const setCategories = jest.fn();
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);

  //Todo lo que se necesita que se reinicialice, que queden con valores sean limpios
  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });

  test("Debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de cambiar la caja de texto", () => {
    const input = wrapper.find("input");
    const value = "Hola";
    input.simulate("change", { target: { value } });
    expect(wrapper.find("p").text().trim()).toBe(value);
  });

  test("NO debe de postear la informaciÃ³n OnSubmit", () => {
    wrapper.find("form").simulate("submit", { preventDefault() {} });

    expect(setCategories).not.toHaveBeenCalled();
  });

  test("Debe de llamar el setCategories y limpiar la caja de texto", () => {
    //Simulando el input change
    const input = wrapper.find("input");
    const value = "Goku";
    input.simulate("change", { target: { value } });

    //Simulando el submit
    wrapper.find("form").simulate("submit", { preventDefault() {} });
    //SetCategories debe de ser llamado
    expect(setCategories).toHaveBeenCalled();
    expect(setCategories).toBeCalledTimes(1);
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

    //El valor del input debe de estar vacio
    expect(input.prop("value")).toBe("");
  });
});
