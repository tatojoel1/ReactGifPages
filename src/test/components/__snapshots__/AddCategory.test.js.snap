// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en el componente AddCategory Debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <FormGroup
    controlId="formBasicEmail"
  >
    <p />
    <FormLabel
      column={false}
      srOnly={false}
    >
      Type in the box to seek some gifs for you
    </FormLabel>
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Example...Tacos"
      type="text"
      value=""
    />
  </FormGroup>
</form>
`;
